### PR TITLE
cortexm_common: Check for the correct runtime CPU during init

### PR DESCRIPTION
**Pull request by gebart**
The added check is meant as a safety measure against flashing an image for the wrong CPU on a board. This is a first basic check which should work for any Cortex-M CPU and checks that the CPU core family is correct. Later we can add another set of vendor specific checks that ensure that the exact CPU model is correct as well.

Running an image for a different CPU will most likely cause a hard fault early on, but in the worst case it may cause damage to hardware since the software will attempt to set hardware registers which may have a completely different function than it expects.
Most CPUs have some way of identifying the exact model as well, but that check is vendor specific and is not included in this first basic check. Most development boards on the other hand, do not have any way of identifying the exact board, other than guessing by probing different pins, which will be unfeasible.

Adds 44-48 bytes of extra ROM usage, no RAM increase.

### CODE REVIEW

**Comment by kaspar030** [comment]
I never accidentally flashed the wrong image to a board using RIOT's build system. How did this happen to you?

**Comment by gebart** [comment]
Having multiple developer boards using the same kind of programmer can lead to mistakes like this. All of the frdm boards for example use the same kind of pin board programmer, I guess the nucleo boards too.

This will be even more hazardous when the OTA firmware updates are working, which is why I think we need to add some more specific platform identification checks after this.

**Comment by gebart** [comment]
The reason I opted for system reset when something is wrong is that a potential OTA bootloader might switch back to a previous image and mark the current image as failed.

File: cpu/cortexm_common/cortexm_init.c [hunk 1]
```diff
@@ -31,8 +31,46 @@
...
+     * binary was compiled for */
+    if (cortexm_check_core() != 0)
+    {
+        /* This binary was built for a different CPU, continuing on will not end
+         * well, so we reset */
+#if DEVELHELP
+        __asm__ volatile ("BKPT #34\n");
+        while(1) {}
+#else
+        NVIC_SystemReset();
```
Comment (kYc0o) [inline cpu/cortexm_common/cortexm_init.c]: You mean we prefer a reboot loop than a crash?

Comment (kYc0o) [reply cpu/cortexm_common/cortexm_init.c]: Oh, sorry, I got it, but only if a bootloader is present to mark the image as failed this is good...

Comment (gebart) [reply cpu/cortexm_common/cortexm_init.c]: A reboot loop is safe. It doesn't mess with any external peripherals, and if you are running on a development board you will see the reset led being lit, it is also possible to halt with the debugger. It is a way better course of action than to keep going and trying to initialize CPU peripherals for a different CPU.

**Comment by kaspar030** [comment]
> This is a first basic check which should work for any Cortex-M CPU and checks that the CPU core family is correct.

> Adds 44-48 bytes of extra ROM usage, no RAM increase.

To me this seems quite expensive given that only the CPU type is checked.
I get 60b extra ROM on samr21-xpro testing with tests/minimal.
Changing the function to "static" saves 16 bytes, using ```while(1){}``` instead of the reset saves another 20b.

> This will be even more hazardous when the OTA firmware updates are working, which is why I think we need to add some more specific platform identification checks after this.

IMO OTA itself should ensure that only correct images are flashed.

**Comment by gebart** [comment]
I think we can close this now. A different approach may be preferable

**Pull request closed without merging**